### PR TITLE
[TR] PIM-5232: Ajaxify product grid family filter

### PR DESCRIPTION
**Pull request by phaseinducer**
| Q                 | A
| ----------------- | ---
| Specs             | yes
| Behats            | n/a
| Blue CI           | yes
| Changelog updated | yes
| Review and 2 GTM  | 

Done in this PR :
- Refactor AjaxChoiceFilter children to avoid overriding the `getForm()` just to change the conf
- Modify the family filter type to be able to filter by codes instead of ids
- Add a new choices filter in front able to fetch choices from any rest route that return a normalized response (PEF style), not select2 formatted one
- Add a formatter that transform the normalized response into the select2 format

### CODE REVIEW

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-choice-filter.js [hunk 1]
```diff
@@ -11,31 +18,7 @@ define(
...
-                            '<% } %>' +
-                            '<input type="text" name="value" value=""/>' +
-                        '</div>' +
-                    '</div>' +
-                    '<div class="btn-group">' +
-                        '<button type="button" class="btn btn-primary filter-update"><%- _.__("Update") %></button>' +
-                    '</div>' +
-                '</div>'
-            ),
+            popupCriteriaTemplate: _.template(template),
```
Comment (jmleroux) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-choice-filter.js]: :+1:

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-rest-choice-filter.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                if (_.isUndefined(this.emptyValue)) {
+                    this.emptyValue = {
+                        type: 'in',
+                        value: ''
+                    };
+                }
+
+                TextFilter.prototype.initialize.apply(this, arguments);
+            },
+
```
Comment (jmleroux) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-rest-choice-filter.js]: Missing documentation. A comment about e.currentTarget could help to understand which div is `$(e.currentTarget).parent().parent().parent()`

Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-rest-choice-filter.js]: I really don't know what all the code here is doing : /
This filter is just a duplication of the select2-choice-filter except it uses the choices formatter and do much less ajax calls (thanks to some dirty override patapouille).

File: src/Pim/Bundle/EnrichBundle/Resources/config/requirejs.yml [hunk 3]
```diff
@@ -202,6 +202,9 @@ config:
         pim/textarea-field:                 pimenrich/js/product/field/textarea-field
         pim/wysiwyg-field:                  pimenrich/js/product/field/wysiwyg-field
 
+        # Fromatters
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Resources/config/requirejs.yml]: `Formatters`

**Comment by jmleroux** [comment]
GTM

File: src/Pim/Bundle/FilterBundle/Filter/AjaxChoiceFilter.php [hunk 4]
```diff
@@ -28,6 +28,18 @@ protected function getFormType()
     /**
      * {@inheritdoc}
      */
+    public function getForm()
+    {
+        if (!$this->form) {
```
Comment (solivier) [inline src/Pim/Bundle/FilterBundle/Filter/AjaxChoiceFilter.php]: ! is for boolean, not sure form here is  a boolean, maybe put "null !== $this->form" ?

**Comment by solivier** [comment]
GTM after comments fixed.

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-rest-choice-filter.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
+'use strict';
```
Comment (juliensnz) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datafilter/filter/select2-rest-choice-filter.js]: love that file name ! :troll:

File: src/Pim/Bundle/DataGridBundle/Resources/public/templates/filter/select2-choice-filter.html [hunk 6]
```diff
@@ -0,0 +1,23 @@
+<div class="choicefilter">
+    <div class="input-prepend">
+        <div class="btn-group">
+            <% if (emptyChoice) { %>
+                <button class="btn dropdown-toggle" data-toggle="dropdown">
```
Comment (skeleton) [inline src/Pim/Bundle/DataGridBundle/Resources/public/templates/filter/select2-choice-filter.html]: can you add a `type` on this button please ?

**Comment by skeleton** [comment]
GTM once fixed.

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php [hunk 7]
```diff
@@ -39,13 +39,18 @@ public function __construct(EntityManagerInterface $entityManager, $entityName)
...
-        $qb->join('f.translations', 'ft');
+        $qb->leftJoin('f.translations', 'ft');
 
-        if (null !== $search) {
+        if (null !== $search && '' !== $search) {
             $qb->where('f.code like :search')->setParameter('search', "%$search%");
             $qb->orWhere('ft.label like :search')->setParameter('search', "%$search%");
         }
 
+        if (isset($options['identifiers']) && is_array($options['identifiers']) && !empty($options['identifiers'])) {
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php]: awesome

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/formatter/choices/base.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+         * [
+         *     { id: 'webcams', text: 'Webcams' },
+         *     { id: 'mugs', text: 'Tassen' }
+         * ]
+         *
+         * @param {Array} entities
+         *
+         * @return {Array}
+         */
+        format: function (entities) {
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/formatter/choices/base.js]: perfect :)

File: src/Pim/Bundle/FilterBundle/Filter/AjaxChoiceFilter.php [hunk 9]
```diff
@@ -80,4 +92,17 @@ protected function parseData($data)
...
+    /**
+     * Return options passed to the form factory
+     *
+     * @return array
+     */
+    protected function getFormOptions()
+    {
+        return array_merge(
+            $this->getOr(FilterUtility::FORM_OPTIONS_KEY, []),
+            ['csrf_protection' => false]
```
Comment (juliensnz) [inline src/Pim/Bundle/FilterBundle/Filter/AjaxChoiceFilter.php]: we need to find a solution about that (maybe not now but it's important)

**Comment by juliensnz** [comment]
Awesome PR dude, GTM after comment fixes

File: src/Pim/Bundle/FilterBundle/Filter/AjaxChoiceFilter.php (hunk 9 above)
Comment (phaseinducer) [inline src/Pim/Bundle/FilterBundle/Filter/AjaxChoiceFilter.php]: Yep, first step would be to understand why we do that. : )

**Pull request merged**